### PR TITLE
Memoization of configs in specific hooks

### PR DESCRIPTION
**Pull request by thephoenixofthevoid**
Please, test this a lot in real applications (not unit tests) before merging. 
In particular, performance implications and correctness.

The commits are quite simple.

There are 2 dependencies added -- they are well known and streamlined. Consider adding them as a devdeps and bundle at the build step.

### CODE REVIEW

**Comment by thephoenixofthevoid** [comment]
In the last commit I copy pasted all dependencies into codebase, dropping irrelevant parts.

**Comment by dbismut** [comment]
Yes!! Thanks! Again something I wanted to do for a while. Indeed I’ll test this with all examples and this will probably ˋ7.1`. I’m not sure how to test performance accurately, if you have any hint let me know. 

When you say include it into devdeps, I’m not sure I follow, I believe tsdx would Bundle them only if they’re included as regular deps no?

**Comment by dbismut** [comment]
Sorry I think our comments crossed each other. All clear.

**Comment by dbismut** [comment]
Thanks again @thephoenixofthevoid, truly appreciate this, never had the chance to have anyone review and refactor my code before. I’m learning a lot.

**Comment by thephoenixofthevoid** [comment]
Please take a look at new changes

**Comment by thephoenixofthevoid** [comment]
It seems like there is a bug in use-case when domTarget is set-up. Please, follow my reasoning and check it:
1. `getBind` checks if `domTarget` is set. We are interested in case when it is.
2. If so, it does some processing and then returns cleanup function.
3. In useRecognizers we make this return function to called twice: 
-  In `Run controller clean functions on unmount`
-  In user invoked `useEffect(bind, [bind])`
I propose to remove the need for `useEffect(bind, [bind])` as this is not needed in any way.
The only issue that may arise is not allowing to pass args, but (!) we may introduce a new option for that case.

**Comment by dbismut** [comment]
@thephoenixofthevoid yes, I know about that one, and thought about removing it for a while. I don't think it's a bug but for sure `clean` is redundant in that situation.

It just feels weird to just use the hook without needing to return anything. Like this right:

```jsx
// before
const bind = useDrag(() => {}, { domTarget })
useEffect(bind, [bind])

// after
useDrag(() => {}, { domTarget })
```

We might try to remove the use of creating the ref separately by having the lib return a ref.

```jsx
const ref = useDrag(() => {}, { ref: true })
// ...
<div ref={ref} />
```

This wouldn't solve the case when `domTarget` is the window though.

**Comment by thephoenixofthevoid** [comment]
Consider useEffect/useLayoutEffect hooks from the React default set of hooks --- they just return nothing.

**Comment by thephoenixofthevoid** [comment]
It's a bug, but it doesn't reveal itself in form of errors because no feature yet depends on calling effect functions and cleaning functions exactly the same number of times. But as we try to optimize further it will manifest for sure.

**Comment by thephoenixofthevoid** [comment]
Speaking about returning a ref from hook: I think we should follow `single responsibility principle` and avoid adding logic that is already handled by useRef hook.

**Comment by dbismut** [comment]
> Consider useEffect/useLayoutEffect hooks from the React default set of hooks --- they just return nothing.

That's true :)

> Speaking about returning a ref from hook: I think we should follow single responsibility principle and avoid adding logic that is already handled by useRef hook.

[Some libs do that](https://github.com/react-spring/react-use-measure), but considering the fact that we allow passing a raw `EventTarget` that's fine!

**Comment by thephoenixofthevoid** [comment]
1. How is testing going? Anything is broken so far?
2. What is the logic behind matchKeysFromObject?
3. What do you think about proposed change? Can we drop the depreciation warning right now?

**Comment by tim-soft** [comment]
@thephoenixofthevoid if you'd like to test this out in a library, I'm using this pretty heavily in https://github.com/tim-soft/react-spring-lightbox#react-spring-lightbox

**Comment by dbismut** [comment]
@thephoenixofthevoid I have reviewed it on GitHub but haven’t tested it yet. In any case this is a good opportunity to bundle the examples directly on this repo rather than somewhere else.

There’s some other issues I need to tackle first, and I have a shitload of work atm but l’ll try to find some time this week.

**Comment by thephoenixofthevoid** [comment]
> @thephoenixofthevoid if you'd like to test this out in a library, I'm using this pretty heavily in https://github.com/tim-soft/react-spring-lightbox#react-spring-lightbox

It looks like you are using the only hook that I have not yet touched. :) I decided to refactor helpers and utilities when working on specific hooks, making it easier to deal with the Boss Hook later.

**Comment by thephoenixofthevoid** [comment]
I have an idea how to handle element handlers that changes every rerender without modifying DOM. Basically the idea is following:
```
controller // reference to controller (in the scope of handleEvent)
controller.handlers // Map of Sets of Functions

function handleEvent(event) {
  const element = this  // reference to an element
  const type = event.type
  const handlers = controller.handlers.get(type)
  for (let handler of handlers) handler.call(this, event)
}

// In bindings object set this function instead of actual handler for every handler defined
```
This way we may change controller.handlers without affecting the actual handler that gets bound to an element. Changing of handlers during component update will not trigger mutation of DOM.

**Comment by dbismut** [comment]
> What is the logic behind matchKeysFromObject?

@thephoenixofthevoid I have absolutely no recollection and looking at it, it looks like something I have copy pasted from SO. It might have been something with nested keys, as the API for config changed a lot while I was investigating for v7. If this works in the tests then I don't think there should be any issue!

**Comment by dbismut** [comment]
@thephoenixofthevoid sorry I'm not sure I follow your proposal, can you be a bit more specific in the situation you're trying to solve? Sorry I'm not deep in this code at the moment and I'm afraid I'm missing the point. What is `element` in this context?

**Comment by thephoenixofthevoid** [comment]
Basically, throw an error if user writes this "useEffect(bind, [bind])", 
handle this logic automatically and, unfortunately, by doing so break backward compatibility.
Also note that bind never changes, so it's the same as "useEffect(bind, [])".

**Comment by thephoenixofthevoid** [comment]
I implemented what I meant mostly in the last commit. I don't have twitter but have telegram.

**Comment by dbismut** [comment]
@thephoenixofthevoid I've done some linting and type fixing, but most importantly injected examples.

You just need to run `yarn install && yarn start` in the `examples` directory and you should have examples running with the library from the source directory with live updates.

Everything runs fine (that's great!), except for the Infinite slideshow example that has a weird behavior (here is how it should behave: https://codesandbox.io/s/infinite-slideshow-nigoy).

But this is most certainly due to react-spring versions being different, I'll have a look at it tomorrow.

File: src/utils/react-fast-compare.ts [hunk 1]
```diff
@@ -1,75 +1,69 @@
...
-      if (!equal(a[keys[i]], b[keys[i]])) return false;
+    for (i = length; i-- !== 0; ) {
+      if (keys[i] === '_owner' && a.$$typeof) continue
+      if (!equal(a[keys[i]], b[keys[i]])) return false
     }
-    return true;
+    return true
   }
 
-  return a !== a && b !== b;
```
Comment (thephoenixofthevoid) [inline src/utils/react-fast-compare.ts]: That's incorrect. It handles NaN case, NaN !== NaN is true, so it says that both a and b should be NaN.

File: src/Controller.ts [hunk 2]
```diff
@@ -22,6 +24,49 @@ type Bindings = Partial<{ [eventName in ReactEventHandlerKey]: Fn[] }>
...
+
+    return bindings
+  }
+
+  public effect = () => {
+    if (this.isDomTargetDefined) {
+      this.bind()
+      this.addDomTargetListeners()
+    }
+    return this!.clean
```
Comment (dbismut) [inline src/Controller.ts:66]: @thephoenixofthevoid 
Why is the `!` needed? Removing it doesn't throw any error?

File: src/hooks/useGesture.ts [hunk 3]
```diff
@@ -1,26 +1,45 @@
...
   InternalHandlers,
   UserHandlers,
-  RecognizerClasses,
   UseGestureConfig,
-  HookReturnType,
 } from '../types'
+import { buildComplexConfig } from './buildConfig'
+
+export function wrapStart(fn: Function) {
+  return function (this: any, { first }: any) {
```
Comment (YChebotaev) [inline src/hooks/useGesture.ts:14]: Don't use `this` as parameter name.

File: src/hooks/useGesture.ts [hunk 4]
```diff
@@ -1,26 +1,45 @@
...
+  }
+}
+
+export function wrapEnd(fn: Function) {
+  return function (this: any, { last }: any) {
+    if (last) fn.apply(this, arguments)
+  }
+}
+
+const RE_NOT_NATIVE = /^on(Drag|Wheel|Scroll|Move|Pinch|Hover)/
```
Comment (YChebotaev) [inline src/hooks/useGesture.ts:25]: Move ALL_UPPERCASE to head of file.

File: src/hooks/useRecognizers.ts [hunk 5]
```diff
@@ -6,67 +6,64 @@ import {
...
+  if (internalHandlers.wheel)   classes.add(WheelRecognizer)
+  if (internalHandlers.scroll)  classes.add(ScrollRecognizer)
+  if (internalHandlers.move)    classes.add(MoveRecognizer)
+  if (internalHandlers.pinch)   classes.add(PinchRecognizer)
+  if (internalHandlers.hover)   classes.add(MoveRecognizer)
 
-  return controller.bind
-}
+  return classes
+}
```
Comment (YChebotaev) [inline src/hooks/useRecognizers.ts:69]: Check newlines as the end of files.

File: src/utils/memoize-one.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    lastResult = resultFn.apply(this, newArgs)
+    calledOnce = true
+    lastThis = this
+    lastArgs = newArgs
+    return lastResult
+  }
+
+  return memoized as ResultFn
```
Comment (YChebotaev) [inline src/utils/memoize-one.ts:28]: Are `as` clause really needed?

File: src/utils/react-fast-compare.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    let length, i, keys
+    if (Array.isArray(a)) {
+      length = a.length
+      if (length !== b.length) return false
+      for (i = length; i-- !== 0; ) if (!equal(a[i], b[i])) return false
+      return true
+    }
+
+    let it
```
Comment (YChebotaev) [inline src/utils/react-fast-compare.ts:21]: Move all variable definitions without initializers to top.

File: src/utils/utils.ts [hunk 8]
```diff
@@ -1,23 +1,66 @@
...
-  return typeof v === 'function' ? v() : v
+/**
+ * Resolves getters (functions) by calling them
+ * If simple value is given it just pasees through
+ *
+ * @param v
+ */
+export function valueFn<T>(v: T | (() => T)): T {
+  if (typeof v === 'function') {
+    // @ts-ignore
```
Comment (YChebotaev) [inline src/utils/utils.ts:61]: Specify exactly error to ignore here.

File: test/Api.test.tsx [hunk 9]
```diff
@@ -4,33 +4,50 @@ import '@testing-library/jest-dom/extend-expect'
 import { BindProps, GenuineHandlers } from './components/Api'
 import Interactive from './components/Interactive'
 import InteractiveDom from './components/InteractiveDom'
+// import InteractiveDomBackwardCompat from './components/InteractiveDomBackwardCompat'
```
Comment (YChebotaev) [inline test/Api.test.tsx:7]: Really needed?

**Pull request merged**